### ISSUE TITLE
Production deployment question

### ISSUE THREAD

**Issue by advance512** [NSLS/universal-webpack-boilerplate#4]
Hi there,

I've noticed that the production server is run like this:

`node ./src/core/server/start.js`

Which means it includes `babel-register` and requires `config.js`. 

Question 1: Is this okay to run in production? I thought the idea was to use webpack, to transpile the code and not need babel-register. Am I correct in saying that `start.js` is development only?

Also, `start.js` uses the `server` function from `universal-webpack`. (Obviously, it needs to call the exported function in `entry.js`, which received the `chunks()` etc.)

In `universal-webpack-settings.js` the input is not `start.js` but rather `entry.js`. This makes sense, since `entry.js` contains the function that `server` needs. Then, the output file is:
`./webroot/build/server/server.js`

The [universal-webpack](https://github.com/halt-hammerzeit/universal-webpack#universal-webpack-settingsjson) readme says:

> The server-side bundle (`settings.server.output` file) is generated from `settings.server.input` file by Webpack when it's run with the webpack.config.server.babel.js configuration.

So the server-side bundle is indeed generated as `./webroot/build/server/server.js`.

BUT, the server-side bundle is based on `entry.js` and not `start.js`, so it doesn't have the `start.js` usage of the `server` function. Hence, the server-side bundle is useless - it never calls the exported function. It's just loading the exported module. 

Question 2: How is the server-side bundle supposed to work without `start.js`? Do I need to deploy the entire project, including source? I thought the bundle was all that is required.

Thanks!

**Comment by ignatevdev** [comment]
First of all, you mentioned the usage of 'babel-register' in the start.js file. Yes, this is not the best practice to use it here but it has a meaning. The only thing I should do at the moment is to add the "ignore" option to make sure it doesn't catch unnecessary files. When I'll do it, you shouldn't worry about `babel-register`, because it wouldn't try to recompile your bundle or `node_modules` and by itself it doesn't give you a performance penalty, rather a bit higher memory usage, which wouldn't be much of a problem since there will be only couple of files processed by babel when you fire your server, but it allows you to keep all of your sources in ES6, which in my opinion is worth it. 

Your second point is that `server()` function in `src/core/server/config.js` is not using the compiled bundle, but instead uses the module exported from `entry.js`, since `start.js` is not specified in the universal-webpack's config. However, this is not true, the `server()` function will always use the compiled bundle and the configuration is correct.

And finally, even though `start.js` is inside the src folder, it should work just fine without the source files present, but I haven't tested it yet. It probably would be better to move it to the root or inside a `bin` folder and I'll probably do that and add a simple Dockerfile so you could test how your project would run without the sources in production.

In conclusion, thank you for your questions, I'll make couple of changes to the project and leave a note on the results here, but for now I'm closing this issue.

**Comment by advance512** [comment]
Thanks for the response. Actually, I talked to the developer of `universal-webpack` and understood how it should work. I personally have a setup where-in I create a client.bundle.js and server.bundle.js, copy a file called server.production.js to that same directory, and just run that. I can also include the package.json and install the necessary dependencies (not devDependencies) and it works perfectly.

A lot of inspiration was derived from your boilerplate, so thanks.
